restapi/src/main/java/org/cristalise/restapi/ItemRoot.java

```diff
@@ -531,6 +532,9 @@ private String executeUploadJob(ItemProxy item, InputStream file, String postDat
                     new OutcomeAttachment(item.getPath(), thisJob.getSchema().getName(), thisJob.getSchema().getVersion(), -1, MediaType.APPLICATION_OCTET_STREAM, binaryData); 
 
             thisJob.setAttachment(outcomeAttachment);

```

Comment:
This is not straightforward. The `ItemUtils.executeJob()` method invoked from `requestTransition()` depends on the contentType and treats the whole `postData` contents as the XML or JSON data. But in `requestBinaryTransition()` the contentType is not XML nor JSON, it is `MULTIPART_FORM_DATA` and `@FormDataParam` annotations are used on the method parameters to extract the two parts of the multipart. If we merge the two methods, we would have to have a big `if-else` statement wrapping the whole contents anyway, because the parameters have to be handled very differently, so not much code can be shared. We would also lose the processing done automatically for the annotations and would have to manually write the iteration for all the parts of a https://docs.oracle.com/javaee/6/api/javax/mail/Multipart.html , get their ContentType and handle them accordingly. By changing the existing method, there's a much higher chance of regressions, much more code will be needed (to manually handle what is now done via the annotations) and also more testing would be needed. I propose to enter a separate issue for this and once the merge is done, we can deprecate the `requestBinaryTransition()` method.

Reply:
This was already implemented in the old restapi repository. Please check this PR: cristal-ise/restapi#47

Reply:
Please use these examples to implement restapi tests as well:

https://github.com/cristal-ise/cristal-ise/tree/develop/testing/src/main/groovy/org/cristalise/restapi/test

Reply:
@kovax @taurus227 
Merging them doesn't work because if there is a `FormDataParam("file")` the `postData` will be `null`. `FormData` doesn't work with `payload` at the same time. In my changes I added a `FormDataParam("outcome")` to store the outcome if the file upload needed an outcome

Reply:
@kovax @taurus227 
I managed to create tests in dsl `SchemaBuilderDynymicFormSpecs` for the schema generation with `inputType: 'file'`

Reply:
The object in cristal is called OutcomeAttachment, which means upload always done together with post data.

I understand that two method cannot be merged, especially for backward compability.